libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/internal/semanticEquals.kt

```diff
@@ -0,0 +1,204 @@
+  when (this) {
+    is GQLDirectiveDefinition -> {
+      if (other !is GQLDirectiveDefinition) {
+        return false
+      }
+
+      if (locations != other.locations) {
+        return false
+      }
+
+      if (repeatable != other.repeatable) {
+        return false
+      }
+    }
+
+    is GQLInputValueDefinition -> {
+      if (other !is GQLInputValueDefinition) {
+        return false
+      }
+
+      if (!type.semanticEquals(other.type)) {
+        return false
+      }
+
+      if (defaultValue != null && other.defaultValue != null) {
+        if (!defaultValue.semanticEquals(other.defaultValue)) {
+          return false
+        }
+      } else if (defaultValue != null || other.defaultValue != null) {
+        return false
+      }
+    }
+
+    is GQLNonNullType -> {
+      if (other !is GQLNonNullType) {
+        return false
+      }
+    }
+
+    is GQLListType -> {
+      if (other !is GQLListType) {
+        return false
+      }
+    }
+
+    is GQLNamedType -> {
+      if (other !is GQLNamedType) {
+        return false
+      }
+    }
+
+    is GQLNullValue -> {
+      if (other !is GQLNullValue) {
+        return false
+      }
+    }
+
+    is GQLListValue -> {
+      if (other !is GQLListValue) {
+        return false
+      }
+    }
+
+    is GQLObjectValue -> {
+      if (other !is GQLObjectValue) {
+        return false
+      }
+    }
+
+    is GQLStringValue -> {
+      if (other !is GQLStringValue) {
+        return false
+      }
+      if (value != other.value) {
+        return false
+      }
+    }
+
+    is GQLBooleanValue -> {
+      if (other !is GQLBooleanValue) {
+        return false
+      }
+      if (value != other.value) {
+        return false
+      }
+    }
+
+    is GQLIntValue -> {
+      if (other !is GQLIntValue) {
+        return false
+      }
+      if (value != other.value) {
+        return false
+      }
+    }
+
+    is GQLFloatValue -> {
+      if (other !is GQLFloatValue) {
+        return false
+      }
+      if (value != other.value) {
+        return false
+      }
+    }
+
+    is GQLEnumValue -> {
+      if (other !is GQLEnumValue) {
+        return false
+      }
+      if (value != other.value) {
+        return false
+      }
+    }
+
+    is GQLVariableValue -> {
+      if (other !is GQLVariableValue) {
+        return false
+      }
+      if (name != other.name) {
+        return false
+      }
+    }
+
+    is GQLEnumTypeDefinition -> {
+      if (other !is GQLEnumTypeDefinition) {
+        return false
+      }
+      if (name != other.name) {
+        return false
+      }
+    }
+
+    is GQLDirective -> {
+      if (other !is GQLDirective) {
+        return false
+      }
+      if (name != other.name) {
+        return false
+      }
+    }
+
+    is GQLArgument -> {
+      if (other !is GQLArgument) {
+        return false
+      }
+      if (name != other.name) {
+        return false
+      }
+    }
+
+    else -> {}
```

Comment:
We should fail with `TODO()` here in case we ever want to try that on other nodes.